#!/usr/bin/env python3
"""
gRPC Smoke Test - PR-3
–ü—Ä–æ–≤–µ—Ä–∫–∞ –±–∞–∑–æ–≤–æ–π —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏ gRPC —Å–µ—Ä–≤–µ—Ä–∞ –∏ –≤–∞–ª–∏–¥–∞—Ü–∏–∏ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä–æ–≤

–ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ:
    python scripts/grpc_smoke.py [HOST] [PORT]
    
–ü—Ä–∏–º–µ—Ä—ã:
    # Production —Å–µ—Ä–≤–µ—Ä
    python scripts/grpc_smoke.py nexy-server.canadacentral.cloudapp.azure.com 443
    
    # –õ–æ–∫–∞–ª—å–Ω—ã–π —Å–µ—Ä–≤–µ—Ä
    python scripts/grpc_smoke.py localhost 50051

–¢–µ—Å—Ç—ã:
    - InterruptSession: –±–∞–∑–æ–≤—ã–π —Ç–µ—Å—Ç –ø—Ä–µ—Ä—ã–≤–∞–Ω–∏—è —Å–µ—Å—Å–∏–∏
    - StreamAudio: —Ç–µ—Å—Ç —Å –≤–∞–ª–∏–¥–Ω—ã–º–∏ session_id –∏ hardware_id
    - StreamAudio (missing session_id): –ø—Ä–æ–≤–µ—Ä–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏ –æ—Ç—Å—É—Ç—Å—Ç–≤—É—é—â–µ–≥–æ session_id
    - StreamAudio (invalid hardware_id): –ø—Ä–æ–≤–µ—Ä–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏ hardware_id="unknown"
"""

import sys
import os
import asyncio
import logging
from pathlib import Path
from typing import Optional

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ –º–æ–¥—É–ª—è–º
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

try:
    import grpc
    from grpc import aio
    import subprocess
    
    # –ò–º–ø–æ—Ä—Ç protobuf —Ñ–∞–π–ª–æ–≤
    sys.path.insert(0, str(project_root / "modules" / "grpc_service"))
    import streaming_pb2
    import streaming_pb2_grpc
except ImportError as e:
    print(f"‚ùå –û—à–∏–±–∫–∞ –∏–º–ø–æ—Ä—Ç–∞: {e}")
    print("–£–±–µ–¥–∏—Ç–µ—Å—å, —á—Ç–æ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω—ã –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏: pip install grpcio grpcio-tools")
    sys.exit(1)

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class GrpcSmokeTest:
    """Smoke-—Ç–µ—Å—Ç –¥–ª—è gRPC —Å–µ—Ä–≤–µ—Ä–∞"""
    
    def __init__(self, host: str, port: int):
        """
        –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è smoke-—Ç–µ—Å—Ç–∞
        
        Args:
            host: –•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞
            port: –ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞
        """
        self.host = host
        self.port = port
        self.channel = None
        self.stub = None
        
    async def connect(self) -> bool:
        """
        –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É
        
        Returns:
            True –µ—Å–ª–∏ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ —É—Å–ø–µ—à–Ω–æ, False –∏–Ω–∞—á–µ
        """
        try:
            # –§–æ—Ä–º–∏—Ä—É–µ–º –∞–¥—Ä–µ—Å
            address = f"{self.host}:{self.port}"
            
            # –î–ª—è HTTPS (–ø–æ—Ä—Ç 443) –∏—Å–ø–æ–ª—å–∑—É–µ–º secure_channel —Å TLS
            # –í–ê–ñ–ù–û: insecure_channel –ù–ï —Ä–∞–±–æ—Ç–∞–µ—Ç —Å TLS –ø–æ—Ä—Ç–æ–º 443
            if self.port == 443:
                # –î–ª—è self-signed —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–∞ —Å–∫–∞—á–∏–≤–∞–µ–º —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç —Å–µ—Ä–≤–µ—Ä–∞
                import subprocess
                try:
                    result = subprocess.run(
                        ['openssl', 's_client', '-connect', address, '-showcerts'],
                        input=b'', capture_output=True, timeout=5
                    )
                    cert_start = result.stdout.find(b'-----BEGIN CERTIFICATE-----')
                    cert_end = result.stdout.find(b'-----END CERTIFICATE-----', cert_start)
                    if cert_start != -1 and cert_end != -1:
                        cert_pem = result.stdout[cert_start:cert_end + len(b'-----END CERTIFICATE-----')]
                        credentials = grpc.ssl_channel_credentials(root_certificates=cert_pem)
                        self.channel = aio.secure_channel(address, credentials)
                        logger.info("‚úÖ –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è secure_channel —Å —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–æ–º —Å–µ—Ä–≤–µ—Ä–∞")
                    else:
                        credentials = grpc.ssl_channel_credentials()
                        self.channel = aio.secure_channel(address, credentials)
                        logger.warning("‚ö†Ô∏è –ò—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è secure_channel –±–µ–∑ —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–∞ (–º–æ–∂–µ—Ç –Ω–µ —Ä–∞–±–æ—Ç–∞—Ç—å)")
                except Exception as e:
                    logger.error(f"–û—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∏—è —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ç–∞: {e}, –∏—Å–ø–æ–ª—å–∑—É–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–µ credentials")
                    credentials = grpc.ssl_channel_credentials()
                    self.channel = aio.secure_channel(address, credentials)
            else:
                # –î–ª—è –ª–æ–∫–∞–ª—å–Ω–æ–≥–æ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è –∏—Å–ø–æ–ª—å–∑—É–µ–º insecure channel
                self.channel = aio.insecure_channel(address)
            
            # –°–æ–∑–¥–∞–µ–º stub
            self.stub = streaming_pb2_grpc.StreamingServiceStub(self.channel)
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ (–ø–æ–ø—ã—Ç–∫–∞ –ø–æ–ª—É—á–∏—Ç—å –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ)
            try:
                await asyncio.wait_for(
                    self.channel.channel_ready(),
                    timeout=5.0
                )
                logger.info(f"‚úÖ –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ {address} —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ")
                return True
            except asyncio.TimeoutError:
                logger.error(f"‚ùå –¢–∞–π–º–∞—É—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ {address}")
                return False
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ {address}: {e}")
                return False
                
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ —Å–æ–∑–¥–∞–Ω–∏—è –∫–∞–Ω–∞–ª–∞: {e}")
            return False
    
    async def test_interrupt_session(self) -> bool:
        """
        –¢–µ—Å—Ç RPC –º–µ—Ç–æ–¥–∞ InterruptSession
        
        Returns:
            True –µ—Å–ª–∏ —Ç–µ—Å—Ç –ø—Ä–æ—à—ë–ª —É—Å–ø–µ—à–Ω–æ, False –∏–Ω–∞—á–µ
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ InterruptSession...")
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å
            request = streaming_pb2.InterruptRequest(
                hardware_id="smoke_test_hardware_id_interrupt"
            )
            
            # –í—ã–∑—ã–≤–∞–µ–º RPC
            response = await self.stub.InterruptSession(
                request,
                timeout=5.0
            )
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º –æ—Ç–≤–µ—Ç
            if response is not None:
                logger.info(f"‚úÖ InterruptSession —É—Å–ø–µ—à–µ–Ω: success={response.success}")
                return True
            else:
                logger.error("‚ùå InterruptSession –≤–µ—Ä–Ω—É–ª None")
                return False
                
        except grpc.RpcError as e:
            logger.error(f"‚ùå gRPC –æ—à–∏–±–∫–∞ –≤ InterruptSession: {e.code()} - {e.details()}")
            return False
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ InterruptSession: {e}")
            return False
    
    async def test_stream_audio(self) -> bool:
        """
        –¢–µ—Å—Ç RPC –º–µ—Ç–æ–¥–∞ StreamAudio (–±–∞–∑–æ–≤—ã–π –≤—ã–∑–æ–≤ —Å –≤–∞–ª–∏–¥–Ω—ã–º–∏ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä–∞–º–∏)
        
        Returns:
            True –µ—Å–ª–∏ —Ç–µ—Å—Ç –ø—Ä–æ—à—ë–ª —É—Å–ø–µ—à–Ω–æ, False –∏–Ω–∞—á–µ
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ StreamAudio —Å –≤–∞–ª–∏–¥–Ω—ã–º–∏ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä–∞–º–∏...")
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å —Å –≤–∞–ª–∏–¥–Ω—ã–º–∏ session_id –∏ hardware_id
            import uuid
            request = streaming_pb2.StreamRequest(
                prompt="test",
                hardware_id="smoke_test_hardware_id",
                session_id=str(uuid.uuid4())
            )
            
            # –í—ã–∑—ã–≤–∞–µ–º RPC (streaming)
            response_count = 0
            async for response in self.stub.StreamAudio(request, timeout=10.0):
                response_count += 1
                
                # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –æ—Ç–≤–µ—Ç –≤–∞–ª–∏–¥–µ–Ω
                if response.WhichOneof("content") is None:
                    logger.warning("‚ö†Ô∏è –ü–æ–ª—É—á–µ–Ω –æ—Ç–≤–µ—Ç –±–µ–∑ content")
                else:
                    content_type = response.WhichOneof("content")
                    logger.info(f"   ‚Üí –ü–æ–ª—É—á–µ–Ω –æ—Ç–≤–µ—Ç: {content_type}")
                    if content_type == "error_message":
                        logger.error(f"   ‚Üí –°–æ–æ–±—â–µ–Ω–∏–µ –æ–± –æ—à–∏–±–∫–µ: {response.error_message}")
                
                # –û–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç–≤–µ—Ç–æ–≤ –¥–ª—è smoke-—Ç–µ—Å—Ç–∞
                if response_count >= 3:
                    break
            
            if response_count > 0:
                logger.info(f"‚úÖ StreamAudio —É—Å–ø–µ—à–µ–Ω: –ø–æ–ª—É—á–µ–Ω–æ {response_count} –æ—Ç–≤–µ—Ç–æ–≤")
                return True
            else:
                logger.error("‚ùå StreamAudio –Ω–µ –≤–µ—Ä–Ω—É–ª –æ—Ç–≤–µ—Ç–æ–≤")
                return False
                
        except grpc.RpcError as e:
            logger.error(f"‚ùå gRPC –æ—à–∏–±–∫–∞ –≤ StreamAudio: {e.code()} - {e.details()}")
            return False
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ StreamAudio: {e}")
            return False
    
    async def test_stream_audio_missing_session_id(self) -> bool:
        """
        –¢–µ—Å—Ç –≤–∞–ª–∏–¥–∞—Ü–∏–∏: –∑–∞–ø—Ä–æ—Å –±–µ–∑ session_id –¥–æ–ª–∂–µ–Ω –≤–µ—Ä–Ω—É—Ç—å INVALID_ARGUMENT
        
        Returns:
            True –µ—Å–ª–∏ —Ç–µ—Å—Ç –ø—Ä–æ—à—ë–ª (–æ—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∞ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ), False –∏–Ω–∞—á–µ
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –≤–∞–ª–∏–¥–∞—Ü–∏–∏: –∑–∞–ø—Ä–æ—Å –±–µ–∑ session_id...")
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å –ë–ï–ó session_id
            request = streaming_pb2.StreamRequest(
                prompt="test",
                hardware_id="smoke_test_hardware_id"
                # session_id –æ—Ç—Å—É—Ç—Å—Ç–≤—É–µ—Ç
            )
            
            # –í—ã–∑—ã–≤–∞–µ–º RPC (streaming)
            error_received = False
            error_message = None
            async for response in self.stub.StreamAudio(request, timeout=10.0):
                content_type = response.WhichOneof("content")
                if content_type == "error_message":
                    error_received = True
                    error_message = response.error_message
                    logger.info(f"   ‚Üí –ü–æ–ª—É—á–µ–Ω–∞ –æ—à–∏–±–∫–∞: {error_message}")
                    break
            
            if error_received and "session_id is required" in error_message.lower():
                logger.info("‚úÖ –í–∞–ª–∏–¥–∞—Ü–∏—è session_id —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω–∞ –æ–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞")
                return True
            else:
                logger.error(f"‚ùå –í–∞–ª–∏–¥–∞—Ü–∏—è session_id –Ω–µ —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω–æ {error_message}")
                return False
                
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                logger.info(f"‚úÖ –í–∞–ª–∏–¥–∞—Ü–∏—è session_id —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω INVALID_ARGUMENT - {e.details()}")
                return True
            else:
                logger.error(f"‚ùå –ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {e.code()} - {e.details()}")
                return False
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ —Ç–µ—Å—Ç–µ –≤–∞–ª–∏–¥–∞—Ü–∏–∏ session_id: {e}")
            return False
    
    async def test_stream_audio_invalid_hardware_id(self) -> bool:
        """
        –¢–µ—Å—Ç –≤–∞–ª–∏–¥–∞—Ü–∏–∏: –∑–∞–ø—Ä–æ—Å —Å hardware_id="unknown" –¥–æ–ª–∂–µ–Ω –≤–µ—Ä–Ω—É—Ç—å –æ—à–∏–±–∫—É
        
        Returns:
            True –µ—Å–ª–∏ —Ç–µ—Å—Ç –ø—Ä–æ—à—ë–ª (–æ—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∞ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ), False –∏–Ω–∞—á–µ
        """
        try:
            logger.info("üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –≤–∞–ª–∏–¥–∞—Ü–∏–∏: –∑–∞–ø—Ä–æ—Å —Å hardware_id='unknown'...")
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å —Å –Ω–µ–≤–∞–ª–∏–¥–Ω—ã–º hardware_id
            import uuid
            request = streaming_pb2.StreamRequest(
                prompt="test",
                hardware_id="unknown",
                session_id=str(uuid.uuid4())
            )
            
            # –í—ã–∑—ã–≤–∞–µ–º RPC (streaming)
            error_received = False
            error_message = None
            async for response in self.stub.StreamAudio(request, timeout=10.0):
                content_type = response.WhichOneof("content")
                if content_type == "error_message":
                    error_received = True
                    error_message = response.error_message
                    logger.info(f"   ‚Üí –ü–æ–ª—É—á–µ–Ω–∞ –æ—à–∏–±–∫–∞: {error_message}")
                    break
            
            if error_received and ("hardware_id" in error_message.lower() or "unknown" in error_message.lower()):
                logger.info("‚úÖ –í–∞–ª–∏–¥–∞—Ü–∏—è hardware_id —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω–∞ –æ–∂–∏–¥–∞–µ–º–∞—è –æ—à–∏–±–∫–∞")
                return True
            else:
                logger.error(f"‚ùå –í–∞–ª–∏–¥–∞—Ü–∏—è hardware_id –Ω–µ —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω–æ {error_message}")
                return False
                
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                logger.info(f"‚úÖ –í–∞–ª–∏–¥–∞—Ü–∏—è hardware_id —Ä–∞–±–æ—Ç–∞–µ—Ç: –ø–æ–ª—É—á–µ–Ω INVALID_ARGUMENT - {e.details()}")
                return True
            else:
                logger.error(f"‚ùå –ù–µ–æ–∂–∏–¥–∞–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {e.code()} - {e.details()}")
                return False
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ —Ç–µ—Å—Ç–µ –≤–∞–ª–∏–¥–∞—Ü–∏–∏ hardware_id: {e}")
            return False
    
    async def run_all_tests(self) -> bool:
        """
        –ó–∞–ø—É—Å–∫ –≤—Å–µ—Ö —Ç–µ—Å—Ç–æ–≤
        
        Returns:
            True –µ—Å–ª–∏ –≤—Å–µ —Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏, False –∏–Ω–∞—á–µ
        """
        results = []
        
        # –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ
        if not await self.connect():
            return False
        
        # –¢–µ—Å—Ç InterruptSession
        results.append(await self.test_interrupt_session())
        
        # –¢–µ—Å—Ç StreamAudio —Å –≤–∞–ª–∏–¥–Ω—ã–º–∏ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä–∞–º–∏
        results.append(await self.test_stream_audio())
        
        # –¢–µ—Å—Ç—ã –≤–∞–ª–∏–¥–∞—Ü–∏–∏ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä–æ–≤
        results.append(await self.test_stream_audio_missing_session_id())
        results.append(await self.test_stream_audio_invalid_hardware_id())
        
        # –ó–∞–∫—Ä—ã–≤–∞–µ–º –∫–∞–Ω–∞–ª
        if self.channel:
            await self.channel.close()
        
        # –ò—Ç–æ–≥–æ–≤—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç
        all_passed = all(results)
        if all_passed:
            logger.info("‚úÖ –í—Å–µ smoke-—Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏ —É—Å–ø–µ—à–Ω–æ")
        else:
            logger.error(f"‚ùå –ù–µ–∫–æ—Ç–æ—Ä—ã–µ —Ç–µ—Å—Ç—ã –Ω–µ –ø—Ä–æ—à–ª–∏: {results}")
        
        return all_passed


async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description="gRPC Smoke Test –¥–ª—è –ø—Ä–æ–≤–µ—Ä–∫–∏ —Å–æ–≤–º–µ—Å—Ç–∏–º–æ—Å—Ç–∏ —Å–µ—Ä–≤–µ—Ä–∞"
    )
    parser.add_argument(
        "host",
        nargs="?",
        default="nexy-server.canadacentral.cloudapp.azure.com",
        help="–•–æ—Å—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: nexy-server.canadacentral.cloudapp.azure.com)"
    )
    parser.add_argument(
        "port",
        nargs="?",
        type=int,
        default=50051,
        help="–ü–æ—Ä—Ç —Å–µ—Ä–≤–µ—Ä–∞ (–ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 50051)"
    )
    
    args = parser.parse_args()
    
    print("=" * 60)
    print("gRPC Smoke Test - PR-3")
    print("=" * 60)
    print(f"–•–æ—Å—Ç: {args.host}")
    print(f"–ü–æ—Ä—Ç: {args.port}")
    print("=" * 60)
    print()
    
    # –°–æ–∑–¥–∞–µ–º —Ç–µ—Å—Ç
    test = GrpcSmokeTest(args.host, args.port)
    
    # –ó–∞–ø—É—Å–∫–∞–µ–º —Ç–µ—Å—Ç—ã
    success = await test.run_all_tests()
    
    print()
    print("=" * 60)
    if success:
        print("‚úÖ –í—Å–µ —Ç–µ—Å—Ç—ã –ø—Ä–æ—à–ª–∏ —É—Å–ø–µ—à–Ω–æ")
        sys.exit(0)
    else:
        print("‚ùå –ù–µ–∫–æ—Ç–æ—Ä—ã–µ —Ç–µ—Å—Ç—ã –Ω–µ –ø—Ä–æ—à–ª–∏")
        sys.exit(1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n‚ö†Ô∏è –ü—Ä–µ—Ä–≤–∞–Ω–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–º")
        sys.exit(1)
    except Exception as e:
        print(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

